top/top.1
.ig \" -*-mode:nroff-*-
Copyright (c) 2002-2004, 2009, Apple Computer, Inc.  All rights reserved.

@APPLE_LICENSE_HEADER_START@

The contents of this file constitute Original Code as defined in and
are subject to the Apple Public Source License Version 1.1 (the
"License").  You may not use this file except in compliance with the
License.  Please obtain a copy of the License at
http://www.apple.com/publicsource and read it before using this file.

This Original Code and all software distributed under the License are
distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT.  Please see the
License for the specific language governing rights and limitations
under the License.

@APPLE_LICENSE_HEADER_END@
..
.TH top 1 "top"
.hy 1
.SH NAME
top - display and update sorted information about processes
.SH SYNOPSIS
.TP
.BR top
.RB [ \-a
|
.B \-d
|
.B \-e
|
.B \-c
.IR <mode> ]
.br
.RB [ \-F
| 
.BR \-f ]
.br
.RB [ \-h ]
.br
.RB [ \-i
.IR <interval> ]
.br
.RB [ \-l
.IR <samples> ]
.br
.RB [ \-ncols
.IR <columns> ]
.br
.RB [ \-o
.IR <key> ]
.RB [ \-O
.IR <skey> ]
.br
.RB [ \-R
|
.BR \-r ]
.br
.RB [ \-S ]
.br
.RB [ \-s
.IR <delay> ]
.br
.RB [ \-n 
.IR <nprocs> ]
.br
.RB [ \-stats 
.IR <keys> ]
.br
.RB [ \-pid
.IR <processid> ]
.br
.RB [ \-user
.IR <username> ]
.br
.RB [ \-U
.IR <username> ]
.br
.RB [ \-u ]
.SH DESCRIPTION
The
.B top
program periodically displays a sorted list of system processes.
The default sorting key is pid, but other keys can be used instead.
Various output options are available.
.SH OPTIONS
Command line option specifications are processed from left to right.
Options can be specified more than once.
If conflicting options are specified, later specifications override earlier
ones.
This makes it viable to create a shell alias for
.B top
with preferred defaults specified, then override those preferred defaults as
desired on the command line.
.TP
.B \-a
Equivalent to -c a.
.TP
.BI \-c " " "" <mode>
Set event counting mode to
.IR <mode> .
The supported modes are:
.RS
.TP
.B a
Accumulative mode.
Count events cumulatively, starting at the launch of
.BR top .
Calculate CPU usage and CPU time since the launch of
.BR top .
.TP
.B d
Delta mode.
Count events relative to the previous sample.
Calculate CPU usage since the previous sample.
This mode by default disables the memory object map reporting.
The memory object map reporting may be re-enabled with the -r option or the interactive r command.
.TP
.B e
Absolute mode.
Count events using absolute counters.
.TP
.B n
Non-event mode (default).
Calculate CPU usage since the previous sample.
.RE
.TP
.B \-d
Equivalent to -c d.
.TP
.B \-e
Equivalent to -c e.
.TP
.B \-F
Do not calculate statistics on shared libraries, also known as frameworks.
.TP
.B \-f 
Calculate statistics on shared libraries, also known as frameworks (default).
.TP
.B \-h
Print command line usage information and exit.
.TP
.BI \-i " " "" <interval>
Update framework (-f) info every 
.I <interval> 
samples; see the 
.B PERFORMANCE vs. ACCURACY
section below for more details.
.TP
.BI \-l " " "" <samples>
Use logging mode and display
.I <samples>
samples, even if standard output is a terminal.
0 is treated as infinity.
Rather than redisplaying, output is periodically printed in raw form.
Note that the first sample displayed will have an invalid %CPU displayed
for each process, as it is calculated using the delta between samples.
.TP
.BI \-ncols " " " " <columns>
Display 
.I <columns>
when using logging mode.
The default is infinite.  The number must be > 0 or an error will occur.
.TP
.BI \-n " " "" <nprocs>
Only display up to
.I <nprocs>
processes.
.TP
.BI \-O " " "" <skey>
Use
.I <skey>
as a secondary key when ordering the process display.
See
.B -o
for key names
.RB ( pid
is the default).
.TP
.BI \-o " " "" <key>
.RS
Order the process display by sorting on
.I <key>
in descending order.
A
.B +
or
.B -
can be prefixed to the key name to specify ascending or descending order,
respectively.
The supported keys are:
.TP
.B pid
Process ID (default).
.TP
.B command
Command name.
.TP
.B cpu
CPU usage.
.TP
.B csw
Number of context switches.
.TP
.B time
Execution time.
.TP
.B threads
alias:
.B th
.br
Number of threads (total/running).
.TP
.B ports
alias:
.B prt
.br
Number of Mach ports.
.TP
.B mregion
alias:
.B mreg, reg
.br
Number of memory regions.
.TP
.B rprvt
Resident private address space size.
.TP
.B rshrd
Resident shared address space size.
.TP
.B rsize
Resident memory size.
.TP
.B vsize
Total memory size.
.TP
.B vprvt
Private address space size.
.TP
.B pgrp
Process group id.
.TP
.B ppid
Parent process id.
.TP
.B state 
alias:
.B pstate
.br
Process state.
.TP
.B uid
User ID.
.TP 
.B wq 
alias:
.B #wq, workqueue
.br
The workqueue total/running.
.TP 
.B faults 
alias:
.B fault
.br
The number of page faults.
.TP
.B cow
alias:
.B cow_faults
.br
The copy-on-write faults.
.TP
.B user
alias:
.B username
Username.
.TP
.B msgsent
.br
Total number of mach messages sent.
.TP 
.B msgrecv
.br
Total number of mach messages received.
.TP
.B sysbsd
Total BSD syscalls.
.TP
.B sysmach
Total Mach syscalls.
.TP
.B pageins
Total pageins.
.RE
.TP
.BI \-R " " ""
Do not traverse and report the memory object map for each process.
.TP
.BI \-r " " ""
Traverse and report the memory object map for each process (default).
.TP
.BI \-S " " ""
Display the global statistics for swap and purgeable memory.
.TP
.BI \-s " " "" <delay>
Set the delay between updates to
.I <delay>
seconds.
The default delay between updates is 1 second.
.TP
.BI \-stats " " "" <keys>
Only display the comma separated statistics.  See the -o flag for the valid
.IR <keys> .
.TP
.BI \-pid " " "" <processid>
Only display 
.IR <processid>
in top.
.TP
.BI \-user " " "" <user>
Only display processes owned by
.IR <user> .
.TP
.BI \-U " " "" <user>
This is an alias for -user.
.TP
.BI \-u
This is an alias equivalent to: -o cpu -O time.

.SH DISPLAY
The first several lines of the
.B top
display show various global state.
All of the information is labeled.
Following is an alphabetical list of global state fields and their descriptions.
.TP 12
.B CPU
Percentage of processor usage, broken into user, system, and idle components.
The time period for which these percentages are calculated depends on the event
counting mode.
.TP 12
.B Disks
Number and total size of disk reads and writes.
.TP 12
.B LoadAvg
Load average over 1, 5, and 15 minutes.
The load average is the average number of jobs in the run queue.
.TP 12
.B MemRegions
Number and total size of memory regions, and total size of memory regions broken
into private (broken into non-library and library) and shared components.
.TP 12
.B Networks
Number and total size of input and output network packets.
.TP 12
.B PhysMem
Physical memory usage, broken into wired, active, inactive, used, and free
components.
.TP 12
.B Procs
Total number of processes and number of processes in each process state.
.TP 12
.B SharedLibs
Resident sizes of code and data segments, and link editor memory usage.
.TP 12
.B Threads
Number of threads.
.TP 12
.B Time
Time, in H:MM:SS format.
When running in logging mode Time is in YYYY/MM/DD HH:MM:SS format by default, but may be overridden with accumulative mode.
When running in accumulative event counting mode, the Time is in HH:MM:SS since the beginning of the top process.
.TP 12
.B VirtMem
Total virtual memory, virtual memory consumed by shared libraries, and number of
pageins and pageouts.
.TP 12
.B Swap
Swap usage: total size of swap areas, amount of swap space in use and amount
of swap space available.
.TP 12
.B Purgeable
Number of pages purged and number of pages currently purgeable.
.PP
Below the global state fields, a list of processes is displayed.
The fields that are displayed depend on the options that are set.
The pid field displays the following for the architecture:
.TP 14
.B +
for 64-bit native architecture, or
.B -
for 32-bit native architecture, or
.B *
for a non-native architecture.
.TP 14
.SH INTERACTION
When
.B top
is run in interactive (non-logging) mode, it is possible to control the output of
.BR top ,
as well as interactively send signals to processes.
The interactive command syntax is terse.
Each command is one character, followed by 0 to 2 arguments.
Commands that take arguments prompt interactively for the arguments, and where
applicable, the default value is shown in square brackets.
The default value can be selected by leaving the input field blank and pressing
enter.
.B ^G
escapes the interactive argument prompt, and has the same effect as leaving
the input field blank and pressing enter.
.PP
The following commands are supported:
.TP
.BR ?
Display the help screen.
Any character exits help screen mode.
This command always works, even in the middle of a command.
.TP
.B ^L
Redraw the screen.
.TP
.BI c <mode>
Set output mode to
.IR <mode> .
The supported modes are:
.RS
.TP
.B a
Accumulative mode.
.TP
.B d
Delta mode.
.TP
.B e
Event mode.
.TP
.B n
Non-event mode.
.RE
.TP
.BI O <skey>
Use
.I <skey>
as a secondary key when ordering the process display.
See the
.B -o
option for key names.
.TP
.BI o <key>
.RS
Order the process display by sorting on
.I <key>
in descending order.
A
.B +
or
.B -
can be prefixed to the key name to specify ascending or descending order,
respectively.
The supported keys and alises are listed with the -o option above.

.RE
.TP
.B q
Quit.
.TP
.B r
Toggle traversal and reporting of the memory object map for each process.
.TP
.BI S <signal> "" <pid>
Send
.I <sig>
to
.IR <pid>.
.I <sig>
can be specified either as a number or as a name (for example,
.BR HUP ).
The default signal starts out as
.BR TERM .
Each time a signal is successfully sent, the default signal is updated to be
that signal.
.I <pid>
is a process id.
.TP
.BI s <delay>
Set the delay between updates to
.I <delay>
seconds.
.TP
.BI U <user>
Only display processes owned by
.IR <user> .
Either the username or uid number can be specified.
To display all processes, press enter without entering a username or uid number.
.SH PERFORMANCE vs. ACCURACY
Calculating detailed memory statistics is fundamentally resource-intensive.
To reduce the cpu usage in top, the
.I \-i
parameter has been introduced to allow the user to tune this tradeoff.  With the 
default value of 10, framework stats will be updated once every 10 samples.
Specifying
.I \-i
1 will result in the most accurate display, at the expense of system resources.
.SH N/A - Not Available
When this occurs in a stat it's caused by the memory object map reporting being 
disabled.  Memory object map reporting is disabled by default in delta mode, but 
may be optionally enabled via -r or the interactive
.B r 
command.  To enable the -r option use it after any -c mode options.
.SH EXAMPLES
.TP
top -o cpu -O +rsize -s 5 -n 20
Sort the processes according to CPU usage (descending) and resident memory size
(ascending), sample and update the display at 5 second intervals, and limit the
display to 20 processes.
.TP
top -c d
Run top in delta mode.

.TP
top -stats pid,command,cpu,th,pstate,time
Display only the specified statistics, regardless of any growth of the terminal.
If the terminal is too small, only the statistics that fit will be displayed.

.SH SEE ALSO
kill(2),
vm_stat(1),
signal(3),
vmmap(1)


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


top/top.c
/*
 * Copyright (c) 2008 Apple Computer, Inc.  All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * The contents of this file constitute Original Code as defined in and
 * are subject to the Apple Public Source License Version 1.1 (the
 * "License").  You may not use this file except in compliance with the
 * License.  Please obtain a copy of the License at
 * http://www.apple.com/publicsource and read it before using this file.
 * 
 * This Original Code and all software distributed under the License are
 * distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT.  Please see the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curses.h>
#include "top.h"
#include "libtop.h"
#include "globalstats.h"
#include "layout.h"
#include "log.h"
#include "userinput.h"
#include "preferences.h"

const libtop_tsamp_t *tsamp;

static int sort_subcomp(int a_key, const libtop_psamp_t *a_a, 
			const libtop_psamp_t *a_b);

static bool top_do_relayout = false;

/* This may not be called in common update patterns. */
void top_relayout_force(void) {
    top_do_relayout = true;
}

void top_relayout(struct statistics_controller *c, int type, int maxwidth) {
    top_do_relayout = true;
}

bool top_need_relayout(void) {
    return top_do_relayout;
}

void *top_create(WINDOW *wmain) {
    void *gstats;
    struct statistics_controller *controller;
    
    gstats = top_globalstats_create(wmain);
    
    if(NULL == gstats) {
	endwin();
	fprintf(stderr, "unable to create global stats!\n");
	_exit(EXIT_FAILURE);
    }

    controller = create_statistics_controller(wmain);
      
    if(NULL == controller) {
	endwin();
	fprintf(stderr, "unable to create controller for main window!\n");
	_exit(EXIT_FAILURE);
    }
    
    controller->globalstats = gstats;

    return controller;
}


static int top_sort(void *a_data, const libtop_psamp_t *a, const libtop_psamp_t *b) {
    int retval;

    retval = sort_subcomp(top_prefs_get_sort(), a, b) * 
	(top_prefs_get_ascending() ? 1 : -1);
    if (retval == 0) {
	retval = sort_subcomp(top_prefs_get_secondary_sort(), a, b) *
	    (top_prefs_get_secondary_ascending() ? 1 : -1);
    }

    return retval;
}

#define COMP(a,b) (((a)==(b)?0:	\
		    ((a)<(b)?-1:1)))

static int sort_subcomp(int a_key, const libtop_psamp_t *a_a, 
		     const libtop_psamp_t *a_b) {
    struct timeval  tv_a, tv_b;
    const char      *user_a, *user_b;
    
    switch (a_key) {
    case STATISTIC_PID: return COMP(a_a->pid, a_b->pid);

    case STATISTIC_COMMAND: return COMP(strcmp(a_a->command, a_b->command),0);

    case STATISTIC_CPU:
	timersub(&a_a->total_time, &a_a->p_total_time, &tv_a);
	timersub(&a_b->total_time, &a_b->p_total_time, &tv_b);

	if(tv_a.tv_sec == tv_b.tv_sec) {
	    return COMP(tv_a.tv_usec, tv_b.tv_usec);
	} else {
	    return COMP(tv_a.tv_sec, tv_b.tv_sec);
	}

    case STATISTIC_TIME:
	 tv_a = a_a->total_time;
	 tv_b = a_b->total_time;

	 if(tv_a.tv_sec == tv_b.tv_sec) {
	     return COMP(tv_a.tv_usec, tv_b.tv_usec);
	 } else {
	     return COMP(tv_a.tv_sec, tv_b.tv_sec);
	 }

    case STATISTIC_THREADS: return COMP(a_a->th, a_b->th);

    case STATISTIC_WORKQUEUE: return COMP(a_a->wq_nthreads, a_b->wq_nthreads);

    case STATISTIC_PORTS: return COMP(a_a->prt, a_b->prt);

    case STATISTIC_MREGION: return COMP(a_a->reg, a_b->reg);

    case STATISTIC_RPRVT: return COMP(a_a->rprvt, a_b->rprvt);

    case STATISTIC_RSHRD: return COMP(a_a->rshrd, a_b->rshrd);

    case STATISTIC_RSIZE: return COMP(a_a->rsize, a_b->rsize);

    case STATISTIC_VSIZE: return COMP(a_a->vsize, a_b->vsize);

    case STATISTIC_VPRVT: return COMP(a_a->vprvt, a_b->vprvt);

    case STATISTIC_PGRP: return COMP(a_a->pgrp, a_b->pgrp);

    case STATISTIC_PPID: return COMP(a_a->ppid, a_b->ppid);

    case STATISTIC_PSTATE: {
	const char *a = libtop_state_str(a_a->state);
	const char *b = libtop_state_str(a_b->state);
	
	return COMP(strcmp(a, b), 0);
    }
    case STATISTIC_UID: return COMP(a_a->uid, a_b->uid);
    
    case STATISTIC_FAULTS:
	return COMP(a_a->faults.now, a_b->faults.now);

    case STATISTIC_COW_FAULTS: 
	return COMP(a_a->cow_faults.now, a_b->cow_faults.now);

    case STATISTIC_MESSAGES_SENT:
	return COMP(a_a->messages_sent.now, a_b->messages_sent.now);
	
    case STATISTIC_MESSAGES_RECEIVED:
	return COMP(a_a->messages_recv.now, a_b->messages_recv.now);

    case STATISTIC_SYSBSD:
	return COMP(a_a->syscalls_bsd.now, a_b->syscalls_bsd.now);

    case STATISTIC_SYSMACH:
	return COMP(a_a->syscalls_mach.now, a_b->syscalls_mach.now);
   
    case STATISTIC_CSW:
	return COMP(a_a->csw.now, a_b->csw.now);

    case STATISTIC_PAGEINS:
	return COMP(a_a->pageins.now, a_b->pageins.now);

    case STATISTIC_USER:
	/* Handle the == case first, since it's common. */
	if (a_a->uid == a_b->uid) return 0;
        
	user_a = libtop_username(a_a->uid);
	user_b = libtop_username(a_b->uid);
	
	return COMP(strcmp(user_a, user_b),0);
    }
    
    return 0;
}

#undef COMP

void top_sample(void) {
    if(libtop_sample(/*calculate mreg, vprvt and more see libtop.h*/
		     top_prefs_get_mmr(), 
		     /*framework stats*/
		     top_prefs_get_frameworks())) {
	endwin();
	fprintf(stderr, "error: while gathering a libtop sample.\n"
		"The permissions and/or ownership are incorrect "
		"for this executable, or you are testing without sudo.\n");
	_exit(EXIT_FAILURE);
    }
}


void top_insert(void *ptr) {
    struct statistics_controller *c = ptr;
    const libtop_psamp_t *psample;
    char *user;
    unsigned long uid = 0;
    int nprocs;
    pid_t pid;
    bool have_pid = false;
    
    c->reset_insertion(c);

    top_sample();

    /* 
     * The ordering is important here, because the libtop_psort actually 
     * updates the tsamp->nprocs.
     */
    libtop_psort(top_sort, NULL);

    tsamp = libtop_tsamp();

    if(top_globalstats_update(c->globalstats, tsamp))
	top_log("An error occurred while updating global stats.\n");
  
    /* The user has requested only displaying processes owned by user. */
    user = top_prefs_get_user();

    if(user)
	uid = top_prefs_get_user_uid();
    
    nprocs = top_prefs_get_nprocs();

    if(0 == nprocs)
	return;
    
    have_pid = top_prefs_get_pid(&pid);

    for(psample = libtop_piterate(); psample; psample = libtop_piterate()) {
	if(user && psample->uid != uid)
	    continue;
	
	if(have_pid) {
	    if(pid != psample->pid)
		continue;
	}

	c->insert_sample(c, psample);

	/*
	 * If nprocs is -1 (the default), or otherwise negative, 
	 * then display all.
	 */
	if(nprocs > 0)
	    --nprocs;
	
	if(0 == nprocs) {
	    break;
	}
    }
}


/* Return true if a non-fatal error occurred. */
bool top_layout(void *ptr) {
    struct statistics_controller *c = ptr;
    int lines = LINES, cols = COLS;
    int consumed_height = 0;
    int pstatheight;

    //top_log("%s\n", __func__);

    top_do_relayout = false;

    //fprintf(stderr, "laying out: lines %d cols %d\n", lines, cols);
    //werase(c->parent);

    if(ERR == wresize(c->parent, lines, cols)) {
	top_log("error: wresizing parent!\n");
	return true;
    }

    top_globalstats_reset(c->globalstats);

    if(top_globalstats_resize(c->globalstats, cols, lines, 
			      &consumed_height)) {
	top_log("error: performing global stats resize!\n");
	return true;
    }

    user_input_set_position(consumed_height, 0);    

    pstatheight = lines - consumed_height - 1;
    if(pstatheight <= 0)
	return true;

    //fprintf(stderr, "consumed_height %d\n", consumed_height);

    if(layout_statistics(c, cols, pstatheight,
			 /*y*/ consumed_height + 1)) {
	top_log("error: performing statistic layout!\n");
	return true;
    }

    return false;
}

struct draw_state {
    int xoffset;
};

static bool top_draw_iterator(struct statistic *s, void *ptr) {
    struct draw_state *state = ptr;

    s->callbacks.draw(s, /*x*/ state->xoffset);

    state->xoffset = 1;

    /* more iterations */
    return true;
}

void top_draw(void *ptr) {
    struct statistics_controller *c = ptr;
    struct draw_state state;

    werase(c->parent);
    
    top_globalstats_draw(c->globalstats);

    user_input_draw(ptr, stdscr);

    state.xoffset = 0;

    c->iterate(c, top_draw_iterator, &state);

    /* This moves the insertion cursor to the lower right. */
    wmove(stdscr, LINES - 1, COLS - 1);
    
    update_panels();
    doupdate();
}
